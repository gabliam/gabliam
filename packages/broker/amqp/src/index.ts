import { AmqpPlugin } from './amqp-plugin';

export * from './amqp-connection';
export * from './amqp-manager';
export * from './errors';
export * from './interfaces';
export * from './metadatas';
export * from './queue';
export { AmqpPlugin as default };
