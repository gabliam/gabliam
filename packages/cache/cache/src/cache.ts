export type ConstructableCache = new (name: string, options?: Record<string, unknown>) => Cache;

/**
 * Interface that defines common cache operations.
 *
 * <b>Note:</b> Due to the generic use of caching, it is recommended that
 * implementations allow storage of <tt>null</tt> values (for example to
 * cache methods that return {@code null}).
 */
export interface Cache {
  start(): Promise<void>;

  stop(): Promise<void>;

  /**
   * Return the cache name.
   */
  getName(): string;

  /**
   * Return the underlying native cache provider.
   */
  getNativeCache(): any;

  /**
   * Return the value to which this cache maps the specified key,
   * generically specifying a type that return value will be cast to.
   * <p>Note: This variant of {@code get} does not allow for differentiating
   * between a cached {@code null} value and no cache entry found at all.
   * Use the standard {@link #get(Object)} variant for that purpose instead.
   * @param key the key whose associated value is to be returned
   * @param type the required type of the returned value (may be
   * {@code null} to bypass a type check; in case of a {@code null}
   * value found in the cache, the specified type is irrelevant)
   * @return the value to which this cache maps the specified key
   * (which may be {@code null} itself), or also {@code null} if
   * the cache contains no mapping for this key
   * @throws IllegalStateException if a cache entry has been found
   * but failed to match the specified type
   * @see #get(Object)
   */
  get<T>(key: string): Promise<T | undefined | null>;

  /**
   * Associate the specified value with the specified key in this cache.
   * <p>If the cache previously contained a mapping for this key, the old
   * value is replaced by the specified value.
   * @param key the key with which the specified value is to be associated
   * @param value the value to be associated with the specified key
   */
  put(key: string, value: any): Promise<void>;

  /**
   * Atomically associate the specified value with the specified key in this cache
   * if it is not set already.
   * <p>This is equivalent to:
   * <pre><code>
   * Object existingValue = cache.get(key);
   * if (existingValue == null) {
   *     cache.put(key, value);
   *     return null;
   * } else {
   *     return existingValue;
   * }
   * </code></pre>
   * except that the action is performed atomically. While all out-of-the-box
   * {@link CacheManager} implementations are able to perform the put atomically,
   * the operation may also be implemented in two steps, e.g. with a check for
   * presence and a subsequent put, in a non-atomic way. Check the documentation
   * of the native cache implementation that you are using for more details.
   * @param key the key with which the specified value is to be associated
   * @param value the value to be associated with the specified key
   * @return the value to which this cache maps the specified key (which may be
   * {@code null} itself), or also {@code null} if the cache did not contain any
   * mapping for that key prior to this call. Returning {@code null} is therefore
   * an indicator that the given {@code value} has been associated with the key.
   * @since 4.1
   */
  putIfAbsent<T>(
    key: string,
    value: T | null | undefined
  ): Promise<T | undefined | null>;

  /**
   * Evict the mapping for this key from this cache if it is present.
   * @param key the key whose mapping is to be removed from the cache
   */
  evict(key: string): Promise<void>;

  /**
   * Remove all mappings from the cache.
   */
  clear(): Promise<void>;
}
