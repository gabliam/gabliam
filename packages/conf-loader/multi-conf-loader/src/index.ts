import {
  configResolver,
  LoaderConfigParseError,
  Resolver,
} from '@gabliam/core';
import d from 'debug';
import fs from 'fs';
import yaml from 'js-yaml';
import _ from 'lodash';
import { inspect, promisify } from 'util';

// Promisify
const glob = promisify(require('glob')); // require and no import for typings bug
const readFile = promisify(fs.readFile);

const debug = d('@gabliam/multi-conf-loader');

async function loadConstant(
  folder: string,
  projectName: string,
  resolver: Resolver,
  profile?: string,
) {
  const files: string[] = await glob(`**/constants?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let constants: object = {};

  if (!files || files.length === 0) {
    return constants;
  }

  const defaultProfileFile = files.find((file) => file === `constants.yml`);
  if (defaultProfileFile) {
    constants = _.merge(
      {},
      constants,
      await loadFile(`${folder}/${defaultProfileFile}`, resolver),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `constants-${profile}.yml`,
    );

    if (profileFile) {
      constants = _.merge(
        {},
        constants,
        await loadFile(`${folder}/${profileFile}`, resolver),
      );
    }
  }
  return constants;
}

/**
 * FileLoader
 */
const multiConfLoader = async (
  { folder, projectName }: { folder: string; projectName: string },
  profile?: string,
) => {
  debug('loadConfig', folder);
  const resolver = configResolver(folder);
  const constants = await loadConstant(folder, projectName, resolver, profile);

  // create tag config
  const ConfigYamlType = new yaml.Type('!config', {
    kind: 'scalar',
    construct: function (data) {
      return _.get(constants, data, {});
    },
    instanceOf: Object,
  });

  const GAB_SCHEMA = yaml.Schema.create([ConfigYamlType]);

  const files: string[] = await glob(`**/application?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let config: object = {};

  if (!files || files.length === 0) {
    return config;
  }

  const defaultProfileFile = files.find((file) => file === `application.yml`);
  if (defaultProfileFile) {
    config = _.merge(
      {},
      config,
      await loadConfigProject(
        `${folder}/${defaultProfileFile}`,
        projectName,
        GAB_SCHEMA,
        resolver,
      ),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `application-${profile}.yml`,
    );

    if (profileFile) {
      config = _.merge(
        {},
        config,
        await loadConfigProject(
          `${folder}/${profileFile}`,
          projectName,
          GAB_SCHEMA,
          resolver,
        ),
      );
    }
  }
  debug('loadedConfig', inspect(config, { depth: null }));
  return config;
};

async function loadConfigProject(
  filePath: string,
  projectName: string,
  schema: yaml.Schema,
  resolver: Resolver,
) {
  const data = await readFile(filePath, 'utf8');
  try {
    const vals = yaml.loadAll(data, undefined, { schema }) || [];
    return await resolver(_.find(vals, <any>{ projectName }) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

async function loadFile(filePath: string, resolver: Resolver): Promise<Object> {
  const data = await readFile(filePath, 'utf8');
  try {
    return await resolver(yaml.load(data) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

export default multiConfLoader;
