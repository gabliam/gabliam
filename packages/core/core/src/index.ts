import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
