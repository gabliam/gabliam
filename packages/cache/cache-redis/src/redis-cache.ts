import { Cache } from '@gabliam/cache';
import Redis from 'ioredis';
import { promisify } from 'util';
import { gunzip, gzip } from 'zlib';

const gzipAsync = promisify(gzip);
const gunzipAsync = promisify(gunzip);

export interface RedisCacheOptions {
  mode?: string;

  duration?: number;

  timeout?: number;

  gzipEnabled?: boolean;

  redisOptions?: Redis.RedisOptions;
}

export class RedisCache implements Cache {
  private client!: Redis.Redis;
  private addTimeout: (promiseFn: Promise<any>) => Promise<any>;

  constructor(private name: string, private options: RedisCacheOptions = {}) {
    if (options.timeout) {
      this.addTimeout = (promiseFn: Promise<any>) =>
        Promise.race([
          promiseFn,
          new Promise((resolve, reject) => {
            setTimeout(() => {
              reject('timeout');
            }, options.timeout);
          }),
        ]);
    } else {
      this.addTimeout = (promiseFn: Promise<any>) => promiseFn;
    }
  }

  async start() {
    this.client = new Redis(this.options.redisOptions);
  }

  async stop() {
    this.client.disconnect();
  }

  getName(): string {
    return this.name;
  }
  getNativeCache(): RedisCache {
    return this;
  }

  async get<T>(key: string, addTimeout = true): Promise<T | undefined | null> {
    const get = (async () => {
      if (!(await this.hasKey(key, false))) {
        return undefined;
      }
      const realKey = this.getKey(key);
      try {
        return this.deserialize(await this.client.getBuffer(realKey));
      } catch {
        /* istanbul ignore next */ {
          return Promise.resolve(undefined);
        }
      }
    })();
    if (addTimeout) {
      return this.addTimeout(get);
    }

    return get;
  }
  async put(key: string, value: any, addTimeout = true): Promise<void> {
    const put = (async () => {
      const realKey = this.getKey(key);
      try {
        // why ? for prevent this message :
        // node_redis: Deprecated: The SET command contains a "undefined" argument.
        // This is converted to a "undefined" string now and will return an error from v.3.0 on.
        // Please handle this in your code to make sure everything works as you intended it to.
        if (
          this.options.mode !== undefined &&
          this.options.duration !== undefined
        ) {
          await this.client.set(
            realKey,
            await this.serialize(value),
            this.options.mode,
            this.options.duration,
          );
        } else {
          await this.client.set(realKey, await this.serialize(value));
        }
      } catch {}
    })();
    if (addTimeout) {
      return this.addTimeout(put);
    }

    return put;
  }

  async putIfAbsent<T>(
    key: string,
    value: T | null | undefined,
    addTimeout = true,
  ): Promise<T | undefined | null> {
    const putIfAbsent = (async () => {
      if (!(await this.hasKey(key, false))) {
        await this.put(key, value, false);
        return null;
      }

      return await this.get<T>(key, false);
    })();

    if (addTimeout) {
      return this.addTimeout(putIfAbsent);
    }

    return putIfAbsent;
  }
  async evict(key: string, addTimeout = true): Promise<void> {
    const evict = (async () => {
      try {
        await this.client.del(this.getKey(key));
      } catch {}
    })();

    if (addTimeout) {
      await this.addTimeout(evict);
    } else {
      await evict;
    }
  }

  async clear(addTimeout = true): Promise<void> {
    const clear = (async () => {
      try {
        await this.client.eval(
          `return redis.call('del', unpack(redis.call('keys', ARGV[1])))`,
          0,
          `${this.name}:*`,
        );
      } catch {}
    })();

    if (addTimeout) {
      await this.addTimeout(clear);
    } else {
      await clear;
    }
  }

  private async hasKey(key: string, addTimeout: boolean) {
    const hasKey = (async () => {
      try {
        return (await this.client.exists(this.getKey(key))) === 1;
      } catch {
        /* istanbul ignore next */ {
          return false;
        }
      }
    })();

    if (addTimeout) {
      return this.addTimeout(hasKey);
    }

    return hasKey;
  }

  private getKey(key: string) {
    return `${this.name}:${key}:${this.name}`;
  }

  private async serialize(value: any) {
    try {
      if (this.options.gzipEnabled === true) {
        return <any>await gzipAsync(JSON.stringify(value));
      } else {
        return JSON.stringify(value);
      }
    } catch {
      return undefined;
    }
  }

  private async deserialize(value: any) {
    try {
      if (this.options.gzipEnabled === true) {
        return JSON.parse(<string>await gunzipAsync(value));
      } else {
        return JSON.parse(value.toString());
      }
    } catch {
      return undefined;
    }
  }
}
