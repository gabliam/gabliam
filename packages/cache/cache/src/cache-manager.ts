import { Cache, ConstructableCache } from './cache';

export interface ICacheGroup {
  defaultCache?: ConstructableCache;

  defaultOptionsCache?: object;

  caches: Map<string, Cache>;
}

/**
 * Gabliam central cache manager SPI.
 * Allows for retrieving named {@link Cache} regions.
 */
export interface CacheManager {
  /**
   * Return the cache associated with the given name.
   * @param name the group cache (must not be {@code null})
   * @param name the cache identifier (must not be {@code null})
   * @return the associated cache, or {@code undefined} if none found
   */
  getCache(name: string, groupName?: string): Promise<Cache | undefined>;

  /**
   * Return a collection of the cache names known by this manager.
   * @return the names of all caches known by the cache manager
   */
  getCacheNames(): string[];
}

export type ConstructableCacheManager = new (
  groups: Map<string, ICacheGroup>,
  dynamic: boolean,
  defaultCache: ConstructableCache,
  defaultOptionsCache?: Record<string, unknown>
) => CacheManager;
