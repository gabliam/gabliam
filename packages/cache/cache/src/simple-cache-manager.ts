import _ from 'lodash';
import { ConstructableCache } from './cache';
import { CacheManager, ICacheGroup } from './cache-manager';
import { NoOpCache } from './caches/no-op-cache';

export class SimpleCacheManager implements CacheManager {
  private startedCache: Map<string, boolean> = new Map();
  constructor(
    private group: Map<string, ICacheGroup>,
    private dynamic: boolean,
    private defaultCache: ConstructableCache = NoOpCache,
    private defaultOptionsCache?: object,
  ) {}

  async getCache(name: string, groupName = 'default') {
    if (!this.group.has(groupName) && this.dynamic) {
      this.group.set(groupName, {
        defaultCache: this.defaultCache,
        defaultOptionsCache: this.defaultOptionsCache,
        caches: new Map(),
      });
    }

    const group = this.group.get(groupName);
    if (!group) {
      return group;
    }

    if (!group.caches.has(name) && this.dynamic) {
      group.caches.set(name, this.constructCache(group, name));
    }

    const cache = group.caches.get(name);

    if (!cache) {
      return cache;
    }

    if (!this.startedCache.has(name)) {
      await cache.start();
      this.startedCache.set(name, true);
    }

    return cache;
  }
  getCacheNames() {
    const names: string[] = [];
    for (const [, group] of this.group.entries()) {
      for (const [, cache] of group.caches.entries()) {
        names.push(cache.getName());
      }
    }

    return names;
  }

  private constructCache(group: ICacheGroup, name: string) {
    const defaultCache = group.defaultCache || this.defaultCache;
    const optionsCache = _.merge(
      {},
      this.defaultOptionsCache || {},
      group.defaultOptionsCache || {},
    );
    return new defaultCache(name, optionsCache);
  }
}
