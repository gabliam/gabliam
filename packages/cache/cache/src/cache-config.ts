import { Bean, InjectContainer, PluginConfig, Value, Joi } from '@gabliam/core';
import { CACHE_MANAGER } from './constant';
import { ConstructableCacheManager, ICacheGroup } from './cache-manager';
import { ConstructableCache } from './cache';
import * as d from 'debug';
import { SimpleCacheManager } from './simple-cache-manager';
import {
  CachePgkNotInstalledError,
  CacheManagerPgkNotInstalledError,
} from './error';
import { MemoryCache, NoOpCache } from './caches';
import * as _ from 'lodash';
const debug = d('Gabliam:Plugin:CachePlugin');

export interface PluginConfig {
  cacheManager: string | ConstructableCacheManager;

  groups: { [k: string]: PluginGroupConfig } | undefined;

  dynamic: boolean;

  defaultCache: string | ConstructableCache;

  defaultOptionsCache?: Object;
}

export interface PluginGroupConfig {
  defaultCache?: ConstructableCache;

  defaultOptionsCache?: object;

  caches: { [k: string]: PluginCacheConfig } | undefined;
}

export interface PluginCacheConfig {
  cache?: string | ConstructableCache;

  options?: Object;
}

const stringOrClass = Joi.alternatives().try(Joi.string(), Joi.func());

const cacheValidator = Joi.object({
  cache: stringOrClass,
  options: Joi.object(),
});

const pluginValidator = Joi.object().keys({
  cacheManager: stringOrClass.default('SimpleCacheManager'),
  dynamic: Joi.boolean().default(true),
  groups: Joi.object()
    .pattern(
      /.*/,
      Joi.object({
        defaultCache: stringOrClass,
        defaultOptionsCache: Joi.object(),
        caches: Joi.object().pattern(/.*/, cacheValidator),
      })
    )
    .default(undefined),
  defaultCache: stringOrClass.default('NoOpCache'),
  defaultOptionsCache: Joi.object(),
});

@InjectContainer()
@PluginConfig()
export class CachePluginConfig {
  @Value('application.cacheConfig', pluginValidator)
  cacheConfig: PluginConfig | undefined;

  @Bean(CACHE_MANAGER)
  createCacheManager(): any {
    if (this.cacheConfig === undefined) {
      return this.createDefaultManager();
    }
    const groups = new Map<string, ICacheGroup>();
    if (this.cacheConfig.groups) {
      for (const [
        groupKey,
        { defaultCache, caches, defaultOptionsCache },
      ] of Object.entries(this.cacheConfig.groups)) {
        const groupCache: ICacheGroup = {
          defaultCache,
          defaultOptionsCache,
          caches: new Map(),
        };

        groups.set(groupKey, groupCache);
        if (caches) {
          for (const [cacheKey, { cache, options }] of Object.entries(caches)) {
            const CacheConstruc = this.getCacheConstruct(
              cache || defaultCache || this.cacheConfig.defaultCache
            );
            const cacheOptions = _.merge(
              {},
              this.cacheConfig.defaultOptionsCache || {},
              defaultOptionsCache || {},
              options || {}
            );

            groupCache.caches.set(
              cacheKey,
              new CacheConstruc(cacheKey, cacheOptions)
            );
          }
        }
      }
    }
    const CacheManagerConstruct = this.getCacheManagerConstruct(
      this.cacheConfig.cacheManager
    );

    return new CacheManagerConstruct(
      groups,
      this.cacheConfig.dynamic,
      this.getCacheConstruct(this.cacheConfig.defaultCache),
      this.cacheConfig.defaultOptionsCache
    );
  }

  private getCacheManagerConstruct(
    cacheManager: string | ConstructableCacheManager
  ): ConstructableCacheManager {
    if (typeof cacheManager === 'string') {
      switch (cacheManager) {
        case 'SimpleCacheManager':
          return SimpleCacheManager;
        default:
          try {
            return <ConstructableCacheManager>require(cacheManager).default;
          } catch {
            throw new CacheManagerPgkNotInstalledError(cacheManager);
          }
      }
    }
    return cacheManager;
  }

  private getCacheConstruct(
    cache: string | ConstructableCache
  ): ConstructableCache {
    if (typeof cache === 'string') {
      switch (cache) {
        case 'MemoryCache':
          return MemoryCache;
        case 'NoOpCache':
          return NoOpCache;
        default:
          try {
            return <ConstructableCache>require(cache).default;
          } catch {
            throw new CachePgkNotInstalledError(cache);
          }
      }
    }

    return cache;
  }

  private createDefaultManager() {
    debug('Create Defaut cache Manager');
    return new SimpleCacheManager(new Map(), true);
  }
}
