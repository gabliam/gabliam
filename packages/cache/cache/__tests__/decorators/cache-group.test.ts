import { Bean, Config, Service } from '@gabliam/core';
import { GabliamTest } from '@gabliam/core/src/testing';
import {
  CacheEvict,
  CacheGroup,
  CachePut,
  CACHE_MANAGER,
  MemoryCache,
  SimpleCacheManager,
} from '../../src/index';

let g: GabliamTest;
let cache: SimpleCacheManager;
beforeEach(async () => {
  g = new GabliamTest();

  @Config()
  class CacheConfig {
    @Bean(CACHE_MANAGER)
    createCache() {
      cache = new SimpleCacheManager(new Map(), true, MemoryCache);
      return cache;
    }
  }

  g.addClass(CacheConfig);
});

describe('cache group', () => {
  test('simple cache', async () => {
    @CacheGroup('test')
    @Service()
    class TestService {
      @CachePut('hi')
      async hi(surname: string, name: string) {
        return `hi ${surname} ${name}`;
      }

      @CacheEvict('hi')
      async evicthi(surname: string, name: string) {}
    }
    g.addClass(TestService);
    await g.build();

    const s = g.gab.container.get(TestService);
    expect(await s.hi('test', 'test')).toMatchSnapshot();
    expect(await s.hi('test', 'test')).toMatchSnapshot();
    expect(await cache.getCache('hi', 'test')).toMatchSnapshot();
    expect(await s.evicthi('test', 'test')).toMatchSnapshot();
    expect(await cache.getCache('hi', 'test')).toMatchSnapshot();
  });
});

test('should fail when decorated multiple times with @CacheGroup', () => {
  expect(function () {
    @CacheGroup('test')
    @CacheGroup('test2')
    class TestBean {}

    new TestBean();
  }).toThrowError();
});

test('should fail when cachename is missing', async () => {
  expect(function () {
    @CacheGroup('test')
    class TestBean {
      @CachePut()
      async hi(surname: string, name: string) {
        return `hi ${surname} ${name}`;
      }
    }

    new TestBean();
  }).toThrowError();
});
