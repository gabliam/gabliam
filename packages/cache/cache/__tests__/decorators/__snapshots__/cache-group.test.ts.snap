// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cache group simple cache 1`] = `"hi test test"`;

exports[`cache group simple cache 2`] = `"hi test test"`;

exports[`cache group simple cache 3`] = `
MemoryCache {
  "name": "hi",
  "options": Object {
    "max": 3,
  },
  "store": LRUCache {
    "allowStale": false,
    "disposeAfter": null,
    "disposed": null,
    "fetchMethod": null,
    "free": Stack {
      "heap": Uint8Array [
        0,
        0,
        0,
      ],
      "length": 0,
    },
    "head": 0,
    "initialFill": 1,
    "keyList": Array [
      "testtest",
      null,
      null,
    ],
    "keyMap": Map {
      "testtest" => 0,
    },
    "max": 3,
    "maxSize": 0,
    "next": Uint8Array [
      0,
      0,
      0,
    ],
    "noDeleteOnFetchRejection": false,
    "noDisposeOnSet": false,
    "noUpdateTTL": false,
    "prev": Uint8Array [
      0,
      0,
      0,
    ],
    "size": 1,
    "sizeCalculation": undefined,
    "tail": 0,
    "ttl": 0,
    "ttlAutopurge": false,
    "ttlResolution": 1,
    "updateAgeOnGet": false,
    "updateAgeOnHas": false,
    "valList": Array [
      "hi test test",
      null,
      null,
    ],
  },
}
`;

exports[`cache group simple cache 4`] = `undefined`;

exports[`cache group simple cache 5`] = `
MemoryCache {
  "name": "hi",
  "options": Object {
    "max": 3,
  },
  "store": LRUCache {
    "allowStale": false,
    "calculatedSize": 0,
    "disposeAfter": null,
    "disposed": null,
    "fetchMethod": null,
    "free": Stack {
      "heap": Uint8Array [
        0,
        0,
        0,
      ],
      "length": 0,
    },
    "head": 0,
    "initialFill": 1,
    "keyList": Array [
      null,
      null,
      null,
    ],
    "keyMap": Map {},
    "max": 3,
    "maxSize": 0,
    "next": Uint8Array [
      0,
      0,
      0,
    ],
    "noDeleteOnFetchRejection": false,
    "noDisposeOnSet": false,
    "noUpdateTTL": false,
    "prev": Uint8Array [
      0,
      0,
      0,
    ],
    "size": 0,
    "sizeCalculation": undefined,
    "tail": 0,
    "ttl": 0,
    "ttlAutopurge": false,
    "ttlResolution": 1,
    "updateAgeOnGet": false,
    "updateAgeOnHas": false,
    "valList": Array [
      null,
      null,
      null,
    ],
  },
}
`;
